### DISCUSSION TITLE
Routes transitions don't use transitions?

### DISCUSSION THREAD

**Discussion by fdutey** [TanStack/router#289] [Q&A]
Hi,

Maybe I missed something (very likely) but it seems to me that when we transition from one route to another, the preload is not happening in a react transition (the one you get by using `useTransition`).
This causes the whole app to suspend, if you actually load resources that suspend. The effect is that it triggers `<Suspense>` component somewhere in the tree and yet, disregards the pending element.

Did I miss the trick to make it use react transitions or the router is not made with this design in mind at all (which would discard using relay, since relay preloadedQueries suspend)?

Thank you

**Comment by mudai11** [comment]
I'm curious about this as well. While using lazy loading, can't I suspense the component and add a fallback loader component?

**Comment by boldurean** [comment]
https://twitter.com/tannerlinsley/status/1732474127712481371

I remember Tanner made question pools related to it but I cannot find results

**Comment by tannerlinsley** [comment]
It is happening inside a React transition, but because useSyncExternalStore doesn’t support transitions fully, it deopts to showing the pending fallbacks immediately

**Reply by bpinto** [reply]
Is this still the case? Because that would explain why I'm getting the pending fallbacks immediately even if I wrap my `navigate()` with a `startTransition()`. That said, I'm still exploring so I could as well be doing something wrong.

**Reply by tannerlinsley** [reply]
Precisely. We use our own transition architecture. This may change down the line when react 19 becomes stable.

**Comment by tigawanna** [comment]
any updates on this?

**Reply by schiller-manuel** [reply]
this did not change

**Comment by dburles** [comment]
Is it possible to use `useTransition` hook with routes? I'd like to be able to display loading state in the current view before the route changes.

**Reply by ViewableGravy** [reply]
It sounds like this is not possible until react exposes some way for external stores to hook into the concurrent rendering of react.

**Comment by ViewableGravy** [comment]
Does anyone know any resource from react that would indicate they are working on a way for external stores to hook into reacts concurrent mode? Would be easier to follow any such resource than asking back here every couple months - although from a bit of googling, I can't find anything from the react team to indicate they even care about exposing this functionality to external stores 

(Which is a real shame, because something like a router where you are extremely likely to have a blocking render is a perfect candidate for async rendering between route transitions)

**Reply by damianstasik** [reply]
https://react.dev/blog/2025/04/23/react-labs-view-transitions-activity-and-more#concurrent-stores

**Comment by matclayton** [comment]
I’ve opened [an RFC proposing a Router-owned presented state for suspending React navigations](https://github.com/TanStack/router/discussions/8188). It builds on the limitation described here and covers the urgent-mount and superseded-navigation cases, SSR and hydration trade-offs, and the open design questions from an opt-in prototype. I’d welcome feedback there.
